src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -1899,6 +1914,51 @@ void Session::send_unbind_message()
 }
 
 
+void Session::send_json_error_message()
+{
+    REALM_ASSERT(m_state == Active);
+    REALM_ASSERT(m_ident_message_sent);
+    REALM_ASSERT(!m_unbind_message_sent);
+    REALM_ASSERT(m_error_to_send);
+    REALM_ASSERT(m_client_error);
+
+    ClientProtocol& protocol = m_conn.get_client_protocol();
+    OutputBuffer& out = m_conn.get_output_buffer();
+    session_ident_type session_ident = get_ident();
+    std::error_code ec = m_client_error->code();
+    auto message = m_client_error->what();
+
+    logger.info("Sending: ERROR \"%1\" (error_code=%2, session_ident=%3)", message, ec.value(),
+                session_ident); // Throws
+
+    nlohmann::json error_body_json;
+    error_body_json["message"] = message;
+    protocol.make_json_error_message(out, session_ident, ec.value(), error_body_json.dump()); // Throws
+
+    m_conn.initiate_write_message(out, this); // Throws
+
+    m_error_to_send = false;
+    m_connection_to_close = true;
+
+    // Connection is waiting to be closed
+    enlist_to_send(); // Throws
+}
+
+
+void Session::close_connection()
+{
+    REALM_ASSERT(m_state == Active);
+    REALM_ASSERT(m_ident_message_sent);
+    REALM_ASSERT(!m_unbind_message_sent);
+    REALM_ASSERT(m_connection_to_close);
+    REALM_ASSERT(m_client_error);
+
+    m_conn.close_due_to_protocol_error(m_client_error->code()); // Throws
+    m_connection_to_close = false;
+    m_client_error = util::none;
```

Comment:
do we need to check whether `m_client_error` is not `util::none` when the session gets destroyed and maybe surface it to the user? i'm thinking about what happens if say we hit a download error and then the client gets disconnected before we are able to send the error message to the server. since `m_conn.close_due_to_protocol_error()` is what calls the connection change listener and surfaces errors, we could loose the error there.

Reply:
What you mention could have happened before my changes were implemented. It would be of course nice to have that edge case covered. I will check if it could be a quick win, otherwise I'll open a ticket.

Reply:
Filed RCORE-1240.